tests/core/pyspec/eth2spec/test/merge/block_processing/test_process_execution_payload.py

```diff
@@ -13,31 +18,184 @@ def run_execution_payload_processing(spec, state, execution_payload, valid=True,
+    next_slot(spec, state)
+
+    # execution payload
+    execution_payload = build_empty_execution_payload(spec, state)
+
+    yield from run_execution_payload_processing(spec, state, execution_payload)
+
+
+@with_merge_and_later
+@spec_state_test
+def test_success_first_payload_with_gap_slot(spec, state):
+    # pre-state
+    state = build_state_with_incomplete_transition(spec, state)
+    next_slot(spec, state)
     next_slot(spec, state)
-    assert not spec.is_transition_completed(state)
 
+    # execution payload
     execution_payload = build_empty_execution_payload(spec, state)
 
     yield from run_execution_payload_processing(spec, state, execution_payload)
+
+
+@with_merge_and_later
+@spec_state_test
+def test_success_regular_payload_with_gap_slot(spec, state):
+    # pre-state
+    state = build_state_with_complete_transition(spec, state)
+    next_slot(spec, state)
+    next_slot(spec, state)
+
+    # execution payload
+    execution_payload = build_empty_execution_payload(spec, state)
+
+    yield from run_execution_payload_processing(spec, state, execution_payload)
+
+
+@with_merge_and_later
+@spec_state_test
+def test_bad_execution_first_payload(spec, state):
+    # completely valid payload, but execution itself fails (e.g. block exceeds gas limit)
+
+    # pre-state
+    state = build_state_with_incomplete_transition(spec, state)
+    next_slot(spec, state)
+
+    # execution payload
+    execution_payload = build_empty_execution_payload(spec, state)
+
+    yield from run_execution_payload_processing(spec, state, execution_payload, valid=False, execution_valid=False)
+
+
+@with_merge_and_later
+@spec_state_test
+def test_bad_execution_regular_payload(spec, state):
+    # completely valid payload, but execution itself fails (e.g. block exceeds gas limit)
+
+    # pre-state
+    state = build_state_with_complete_transition(spec, state)
+    next_slot(spec, state)
+
+    # execution payload
+    execution_payload = build_empty_execution_payload(spec, state)
+
+    yield from run_execution_payload_processing(spec, state, execution_payload, valid=False, execution_valid=False)
+
+
+@with_merge_and_later
+@spec_state_test
+def test_bad_parent_hash_regular_payload(spec, state):
+    # pre-state
+    state = build_state_with_complete_transition(spec, state)
+    next_slot(spec, state)
+
+    # execution payload
+    execution_payload = build_empty_execution_payload(spec, state)
+    execution_payload.parent_hash = spec.Hash32()
+
+    yield from run_execution_payload_processing(spec, state, execution_payload, valid=False)
+
+
+@with_merge_and_later
+@spec_state_test
+def test_bad_number_regular_payload(spec, state):
+    # pre-state
+    state = build_state_with_complete_transition(spec, state)
+    next_slot(spec, state)
+
+    # execution payload
+    execution_payload = build_empty_execution_payload(spec, state)
+    execution_payload.number = execution_payload.number + 1
```

Comment:
noting that `execution_payload.number` is already correct from `build_empty_execution_payload` so that the increment makes it _incorrect_

Reply:
I agree, but the naming is this way to match the existing Merge spec: https://github.com/ethereum/rayonism/blob/master/specs/merge.md This is as far as we have for an API for the Merge. I don't think we should change it in this PR. With the API redesign in the coming weeks we can rename it.